packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,159 @@
+    String packageName = mContext.getPackageName();
+
+    constants.put("applicationName", applicationName);
+    constants.put("bundleId", packageName);
+
+    PackageManager packageManager = mContext.getPackageManager();
+    try {
+      PackageInfo pInfo = packageManager.getPackageInfo(packageName, 0);
+      constants.put("nativeAppVersion", pInfo.versionName);
+      constants.put("nativeBuildVersion", pInfo.versionCode);
+    } catch (PackageManager.NameNotFoundException e) {
+      e.printStackTrace();
+    }
+
+    constants.put("androidId", Settings.Secure.getString(mContext.getContentResolver(), Settings.Secure.ANDROID_ID));
+
+    return constants;
+  }
+
+  @ExpoMethod
+  public void getFirstInstallTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.firstInstallTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get first install time of this application.", e);
+    }
+  }
+
+  @ExpoMethod
+  public void getLastUpdateTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get last update time of this application.", e);
+    }
+  }
+
+
+  @ExpoMethod
+  public void getInstallReferrerAsync(final Promise promise) {
+
+    final StringBuilder installReferrer = new StringBuilder();
+
+    final InstallReferrerClient referrerClient;
+    referrerClient = InstallReferrerClient.newBuilder(mContext).build();
+    referrerClient.startConnection(new InstallReferrerStateListener() {
+      @Override
+      public void onInstallReferrerSetupFinished(int responseCode) {
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established
+            Log.d("INSTALL_REFERRER", "connection established");
+            try {
+              ReferrerDetails response = referrerClient.getInstallReferrer();
+              installReferrer.append(response.getInstallReferrer());
+              response.getReferrerClickTimestampSeconds();
+              response.getInstallBeginTimestampSeconds();
+              promise.resolve(installReferrer.toString());
+              referrerClient.endConnection();
+            } catch (RemoteException e) {
+              e.printStackTrace();
+              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Binding to play store remote invocation error.", e);
```

Comment:
This error message is really cryptic.
```suggestion
              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "The Play Store raised an error when retrieving the install referrer", e);
```